packages/front-end/components/MetricAnalysis/HistogramGraph.tsx

```diff
@@ -0,0 +1,260 @@
+    setHoverBin(bin);
+  };
+
+  useEffect(
+    () => {
+      if (hoverBin === null) {
+        hideTooltip();
+        return;
+      }
+      const datapoint = data[hoverBin];
+      if (!datapoint) {
+        hideTooltip();
+        return;
+      }
+      const innerWidth =
+        width - marginLeft - marginRight + width / data.length - 1;
+      const tooltipData = getTooltipDataFromDatapoint(
+        datapoint,
+        data,
+        innerWidth,
+        yScale
+      );
+      if (!tooltipData) {
+        hideTooltip();
+        return;
+      }
+
+      showTooltip({
+        tooltipLeft: tooltipData.x,
+        tooltipTop: tooltipData.y,
+        tooltipData: tooltipData,
+      });
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [hoverBin, data, marginLeft, marginRight, width, showTooltip, hideTooltip]
+  );
+
+  return (
+    <ParentSizeModern style={{ position: "relative" }}>
+      {({ width }) => {
+        const xMax = width - marginRight - marginLeft;
+
+        const handlePointerMove = (
+          event: React.PointerEvent<HTMLDivElement>
+        ) => {
+          // coordinates should be relative to the container in which Tooltip is rendered
+          const containerX =
+            ("clientX" in event ? event.clientX : 0) - containerBounds.left;
+          const bin = Math.min(
+            Math.max(0, Math.floor((data.length * containerX) / xMax)),
+            data.length - 1
+          );
+          onHover(bin);
+        };
+
+        const handlePointerLeave = () => {
+          hideTooltip();
+          onHover(null);
+        };
+
+        return (
+          <>
+            <div
+              ref={containerRef}
+              style={{
+                zIndex: 1,
+                position: "absolute",
+                overflow: "hidden",
+                width: xMax,
+                height: graphHeight,
+                marginLeft: marginLeft,
+                marginTop: marginTop,
+              }}
+              onPointerMove={handlePointerMove}
+              onPointerLeave={handlePointerLeave}
+            >
+              {tooltipOpen && (
+                <>
+                  <div
+                    className={styles.positionIndicator}
+                    style={{
+                      transform: `translate(${tooltipLeft}px, ${tooltipTop}px)`,
+                    }}
+                  />
+                  <TooltipWithBounds
+                    left={tooltipLeft}
+                    top={tooltipTop}
+                    className={styles.tooltip}
+                    unstyled={true}
+                  >
+                    {tooltipData && getTooltipContents(tooltipData, formatter)}
+                  </TooltipWithBounds>
+                </>
+              )}
+            </div>
+            <svg width={width} height={height}>
+              <Group top={marginTop} left={marginLeft}>
+                {binWidth
+                  ? data.map((d, i) => (
```

Comment:
its better to break looping functions outside of the main render. so it would be best to put this data.map into a funtion and then call it from here

Reply:
you can also do the binwidwith logic in that function to tidy up the main render

Reply:
Personal pref, but i disagree. For simple maps over data with simple render logic, this is more readable to me. Here's how AirBnB renders their histograms:
https://airbnb.io/visx/bars